File: cropaway/Models/ClipTransition.swift
//
//  ClipTransition.swift
//  cropaway
//

import Foundation
import Combine

/// Type of transition between two clips in a sequence
enum TransitionType: String, Codable, CaseIterable, Identifiable {
    case cut = "cut"
    case fade = "fade"
    case fadeToBlack = "fadeToBlack"
    case opticalFlow = "opticalFlow"

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .cut: return "Cut"
        case .fade: return "Fade"
        case .fadeToBlack: return "Fade to Black"
        case .opticalFlow: return "Morph"
        }
    }

    var iconName: String {
        switch self {
        case .cut: return "scissors"
        case .fade: return "circle.lefthalf.filled"
        case .fadeToBlack: return "circle.bottomhalf.filled"
        case .opticalFlow: return "wand.and.rays"
        }
    }

    /// Whether this transition requires a duration parameter
    var requiresDuration: Bool {
        self != .cut
    }

    /// Default duration for this transition type
    var defaultDuration: Double {
        switch self {
        case .cut: return 0
        case .fade, .fadeToBlack: return 0.5
        case .opticalFlow: return 0.5
        }
    }

    /// Whether this transition type is available on the current macOS version
    var isAvailable: Bool {
        switch self {
        case .cut, .fade, .fadeToBlack:
            return true
        case .opticalFlow:
            if #available(macOS 26.0, *) {
                return true
            } else {
                return false
            }
        }
    }

    /// Available transition types for the current macOS version
    static var availableTypes: [TransitionType] {
        allCases.filter { $0.isAvailable }
    }
}

/// Represents a transition between two clips in a timeline sequence
final class ClipTransition: Identifiable, ObservableObject, Codable {
    let id: UUID

    /// Type of transition (cut, fade, fade to black, or optical flow morph)
    @Published var type: TransitionType

    /// Duration of the transition in seconds (used for all types except cut)
    /// Range: 0.1 to 2.0 seconds
    @Published var duration: Double

    /// Index of the clip this transition follows (transition occurs after this clip)
    let afterClipIndex: Int

    init(
        id: UUID = UUID(),
        type: TransitionType = .cut,
        duration: Double = 0.5,
        afterClipIndex: Int
    ) {
        self.id = id
        self.type = type
        self.duration = max(0.1, min(2.0, duration))
        self.afterClipIndex = afterClipIndex
    }

    // MARK: - Codable

    enum CodingKeys: String, CodingKey {
        case id, type, duration, afterClipIndex
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(UUID.self, forKey: .id)
        type = try container.decode(TransitionType.self, forKey: .type)
        duration = try container.decode(Double.self, forKey: .duration)
        afterClipIndex = try container.decode(Int.self, forKey: .afterClipIndex)
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(id, forKey: .id)
        try container.encode(type, forKey: .type)
        try container.encode(duration, forKey: .duration)
        try container.encode(afterClipIndex, forKey: .afterClipIndex)
    }

    // MARK: - Computed Properties

    /// Effective duration of this transition (0 for cuts, duration for all others)
    var effectiveDuration: Double {
        type.requiresDuration ? duration : 0
    }

    /// Creates a copy of this transition with a new afterClipIndex
    func copy(withNewIndex newIndex: Int) -> ClipTransition {
        ClipTransition(
            id: UUID(),
            type: type,
            duration: duration,
            afterClipIndex: newIndex
        )
    }
}

File: CropawayWindows/CropawayWindows/Models/ClipTransition.cs
// ClipTransition.cs
// CropawayWindows

using System.Text.Json.Serialization;
using CommunityToolkit.Mvvm.ComponentModel;

namespace CropawayWindows.Models;

/// <summary>
/// Represents a transition between two clips in a timeline sequence.
/// Indexed by the clip it follows (afterClipIndex). Duration is clamped
/// to the 0.1-2.0 second range.
/// </summary>
public partial class ClipTransition : ObservableObject, IEquatable<ClipTransition>
{
    /// <summary>
    /// Unique identifier for this transition.
    /// </summary>
    public Guid Id { get; init; } = Guid.NewGuid();

    /// <summary>
    /// Type of transition effect (Cut, Fade, or FadeToBlack).
    /// </summary>
    [ObservableProperty]
    private TransitionType _type = TransitionType.Cut;

    /// <summary>
    /// Duration of the transition in seconds. Clamped to the range 0.1-2.0.
    /// Ignored for Cut transitions (which are instantaneous).
    /// </summary>
    [ObservableProperty]
    private double _duration = 0.5;

    /// <summary>
    /// Index of the clip this transition follows. The transition occurs
    /// between clip[AfterClipIndex] and clip[AfterClipIndex + 1].
    /// </summary>
    public int AfterClipIndex { get; init; }

    public ClipTransition()
    {
    }

    public ClipTransition(TransitionType type, double duration, int afterClipIndex)
    {
        Type = type;
        _duration = Math.Clamp(duration, 0.1, 2.0);
        AfterClipIndex = afterClipIndex;
    }

    // -- Clamping on property change --

    partial void OnDurationChanged(double value)
    {
        double clamped = Math.Clamp(value, 0.1, 2.0);
        if (Math.Abs(clamped - value) > 0.0001)
        {
            Duration = clamped;
        }
    }

    // -- Computed properties --

    /// <summary>
    /// Effective duration of this transition. Returns 0 for Cut transitions
    /// (which are instantaneous) and the configured duration for all others.
    /// </summary>
    [JsonIgnore]
    public double EffectiveDuration =>
        Type.RequiresDuration() ? Duration : 0;

    /// <summary>
    /// Creates a copy of this transition with a different afterClipIndex.
    /// Used when clips are inserted, removed, or reordered in the timeline.
    /// </summary>
    public ClipTransition CopyWithNewIndex(int newIndex)
    {
        return new ClipTransition(Type, Duration, newIndex);
    }

    public bool Equals(ClipTransition? other)
    {
        if (other is null) return false;
        if (ReferenceEquals(this, other)) return true;
        return Id == other.Id;
    }

    public override bool Equals(object? obj) => Equals(obj as ClipTransition);

    public override int GetHashCode() => Id.GetHashCode();

    public static bool operator ==(ClipTransition? left, ClipTransition? right) =>
        left is null ? right is null : left.Equals(right);

    public static bool operator !=(ClipTransition? left, ClipTransition? right) =>
        !(left == right);

    public override string ToString() =>
        $"ClipTransition({Type.DisplayName()}, {Duration:F2}s, after clip {AfterClipIndex})";
}
